controllers/iamauthpolicy_controller.go

```diff
@@ -29,34 +36,156 @@ func RegisterIAMAuthPolicyController(log gwlog.Logger, mgr ctrl.Manager) error {
+			reconcileFunc = c.deleteRoutePolicy
+		} else {
+			reconcileFunc = c.upsertRoutePolicy
+		}
+	}
+
+	latticeResourceId, err := reconcileFunc(ctx, k8sPolicy)
+	if err != nil {
+		if services.IsNotFoundError(err) {
+			c.log.Infof("reconcile error, retry in 30sec: %s", err)
+			return ctrl.Result{RequeueAfter: time.Second * 30}, nil
+		}
+		return ctrl.Result{}, err
 	}
+
+	k8sPolicy.Annotations["application-networking.k8s.aws/resourceId"] = latticeResourceId
+
+	err = c.client.Update(ctx, k8sPolicy)
 	if err != nil {
 		return ctrl.Result{}, err
 	}
 
-	c.log.Infow("successfully reconciled", "req", req)
+	c.log.Infow("reconciled IAM policy",
+		"req", req,
+		"targetRef", k8sPolicy.Spec.TargetRef,
+		"latticeResorceId", latticeResourceId,
+		"isDeleted", isDelete,
+	)
 	return ctrl.Result{}, nil
 }
 
+func (c *IAMAuthPolicyController) handleFinalizer(ctx context.Context, k8sPolicy *anv1alpha1.IAMAuthPolicy) {
+	authPolicyFinalizer := "iamauthpolicy.k8s.aws/resources"
+	if k8sPolicy.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(k8sPolicy, authPolicyFinalizer) {
+			controllerutil.AddFinalizer(k8sPolicy, authPolicyFinalizer)
+		}
+	} else {
+		if controllerutil.ContainsFinalizer(k8sPolicy, authPolicyFinalizer) {
+			controllerutil.RemoveFinalizer(k8sPolicy, authPolicyFinalizer)
+		}
+	}
+}
+
+func (c *IAMAuthPolicyController) deleteGatewayPolicy(ctx context.Context, k8sPolicy *anv1alpha1.IAMAuthPolicy) (string, error) {
+	snId, err := c.findSnId(ctx, k8sPolicy)
+	if err != nil {
+		return "", err
+	}
+	err = c.policyMgr.Delete(ctx, snId)
+	if err != nil {
+		return "", err
+	}
+	return snId, nil
+}
+
+func (c *IAMAuthPolicyController) findSnId(ctx context.Context, k8sPolicy *anv1alpha1.IAMAuthPolicy) (string, error) {
+	tr := k8sPolicy.Spec.TargetRef
+	snInfo, err := c.policyMgr.Cloud.Lattice().FindServiceNetworkByK8sName(ctx, string(tr.Name))
+	if err != nil {
+		return "", err
+	}
+	return *snInfo.SvcNetwork.Id, nil
+}
+
+func (c *IAMAuthPolicyController) upsertGatewayPolicy(ctx context.Context, k8sPolicy *anv1alpha1.IAMAuthPolicy) (string, error) {
+	snId, err := c.findSnId(ctx, k8sPolicy)
+	if err != nil {
+		return "", err
+	}
+	err = c.policyMgr.EnableSnIAMAuth(ctx, snId)
+	if err != nil {
+		return "", err
+	}
+	err = c.putPolicy(ctx, snId, k8sPolicy.Spec.Policy)
+	if err != nil {
+		return "", err
+	}
+
+	return snId, nil
+}
+
+func (c *IAMAuthPolicyController) findSvcId(ctx context.Context, k8sPolicy *anv1alpha1.IAMAuthPolicy) (string, error) {
+	tr := k8sPolicy.Spec.TargetRef
+	svcName := utils.LatticeServiceName(string(tr.Name), k8sPolicy.Namespace)
+	svcInfo, err := c.policyMgr.Cloud.Lattice().FindServiceByK8sName(ctx, svcName)
+	if err != nil {
+		return "", err
+	}
+	return *svcInfo.Id, nil
+}
+
+func (c *IAMAuthPolicyController) deleteRoutePolicy(ctx context.Context, k8sPolicy *anv1alpha1.IAMAuthPolicy) (string, error) {
+	svcId, err := c.findSvcId(ctx, k8sPolicy)
+	if err != nil {
+		return "", err
+	}
```

Comment:
Do we need to `DisableSvcIAMAuth`  here by updating the latticeService AuthType to "NONE"? (I am not 100% sure on lattice side how this work)


(Same thing for `DisableSnIAMAuth`)

Reply:
Make sense